### PR TITLE
fix(amf): Decoding failure for extended UE Security Capabilities

### LINKED ISSUE #12167: [AMF]: Decoding failure for extended UE Security Capabilities
When UE sends complete 8 octets of UE security capability the decoding fails in NAS5G decoder[https://app.zenhub.com/files/170803235/1ec1f177-3846-4878-86b3-f8d00b5e2875/download](https://app.zenhub.com/files/170803235/1ec1f177-3846-4878-86b3-f8d00b5e2875/download)[https://app.zenhub.com/files/170803235/4639906a-1684-4b06-891b-2686257abdb4/download](https://app.zenhub.com/files/170803235/4639906a-1684-4b06-891b-2686257abdb4/download)

### LINKED ISSUE #12280: [AMF]: Malformed Security Mode Command packet due to UE Security Capability
UE Security Capabilties can have extended set. With that set SMC is getting malformed. Attaching the pcap for reference.

Idea is to handle the UE capabilities in SMC correctly.

[https://app.zenhub.com/files/170803235/b59a2f3e-5a80-475a-b462-ea48fda57621/download]

(https://app.zenhub.com/files/170803235/b59a2f3e-5a80-475a-b462-ea48fda57621/download)

### PR DESCRIPTION
**Pull request by sreedharkumartn**
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

## Summary
Issue:
- When UE sends complete 8 octets of UE security capability the decoding fails in NAS5G decoder.
Issue_Pcap

Fix:
- By adding the logic to skip the non supported bytes.

## Test Plan
- Validated with UT.
UT
- Basic sanity with UERANSIM.
Pcap Snap:
pcap
mme log:
[mme.log](https://github.com/magma/magma/files/8287572/mme.log)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp [hunk 1]
```diff
@@ -99,6 +100,12 @@ int UESecurityCapabilityMsg::DecodeUESecurityCapabilityMsg(
     decoded++;
 
     // decoded = type_len + length_len + ue_sec_capability->length;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp]: remove if not used

Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp [hunk 2]
```diff
@@ -98,7 +99,11 @@ int UESecurityCapabilityMsg::DecodeUESecurityCapabilityMsg(
     ue_sec_capability->eia7 = *(buffer + decoded) & 0x1;
     decoded++;
 
-    // decoded = type_len + length_len + ue_sec_capability->length;
+    // Skipping the remaining bytes as not supported
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp:101]: Statements at https://github.com/magma/magma/pull/12179/files#diff-3a0779f7713eb9082f1c0c9d4bde04e33585ac81524a9c38507d8f3ebaafe3eeR77 can be removed

Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp:101]: Addressed Review Comment

**Comment by rsarwad** [comment]
LGTM

**Review by rsarwad — APPROVED** [review]
LGTM

**Review by ganeshg87 — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp [hunk 3]
```diff
@@ -98,7 +98,11 @@ int UESecurityCapabilityMsg::DecodeUESecurityCapabilityMsg(
...
+    if (ue_sec_capability->length > (decoded - (type_len + length_len))) {
+      decoded +=
+          ue_sec_capability->length - (decoded - (type_len + length_len));
+    }
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp:108]: the fix is needed to encode function also and verify encode and decode functions of  UE security capabilities

Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp:108]: Addressed Review Comment

**Comment by sreedharkumartn** [comment]
Made additional change to fix the issue #12280. Kindly review it.
Done UT testing and Basic sanity with UERANSIM.
UT Test:
UT

Pcap snap:
image

mme log:
[mme.log](https://github.com/magma/magma/files/8380977/mme.log)

**Review by ganeshg87 — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp [hunk 4]
```diff
@@ -98,7 +98,15 @@ int UESecurityCapabilityMsg::DecodeUESecurityCapabilityMsg(
     ue_sec_capability->eia7 = *(buffer + decoded) & 0x1;
     decoded++;
 
-    // decoded = type_len + length_len + ue_sec_capability->length;
+    // Skipping the remaining bytes as not supported
+    if (ue_sec_capability->length > (decoded - (type_len + length_len))) {
+      // The length of the bytes skipped
+      uint8_t add_ue_sec_cap_length =
+          ue_sec_capability->length - (decoded - (type_len + length_len));
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp:105]: Line numbers 105 and 107 are same i.e ue_sec_capability->length - (decoded - (type_len + length_len)). 
decoded += add_ue_sec_cap_length;
ue_sec_capability->length -= add_ue_sec_cap_length;

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp [hunk 5]
```diff
@@ -98,7 +98,15 @@ int UESecurityCapabilityMsg::DecodeUESecurityCapabilityMsg(
     ue_sec_capability->eia7 = *(buffer + decoded) & 0x1;
     decoded++;
 
-    // decoded = type_len + length_len + ue_sec_capability->length;
+    // Skipping the remaining bytes as not supported
+    if (ue_sec_capability->length > (decoded - (type_len + length_len))) {
+      // The length of the bytes skipped
+      uint8_t add_ue_sec_cap_length =
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp]: add_ue_sec_cap_length, is this variable meant to say "additional_ue_sec_cap_length" i.e number of bytes to be skipped. Could you rename the variable.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp (hunk 4 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp:105]: Addressed Review comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp (hunk 5 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GUESecurityCapability.cpp]: Addressed Review Comment

**Pull request merged**